content/docs/user-guide/dvc-files.md

```diff
@@ -237,14 +240,233 @@ stages:
+```
+
+The substitution expression supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the expression literally in `dvc.yaml`, escape it with a backslash,
+> e.g. `\${...`.
+
+### Defining multiple stages at once
+
+You can generate more than one stage from a single `dvc.yaml` entry with the
+following syntax. A `foreach` element accepts a list or dictionary with values
+to iterate on, while `do` contains the regular stage fields (`cmd`, `outs`,
+etc.). Here's a simple example:
+
+```yaml
+stages:
+  echo: # Multi-stage
+    foreach: # List of simple values
+      - foo
+      - bar
+      - baz qux
+    do:
+      cmd: echo "${item}"
+```
+
+Upon `dvc repro`, each item in the list is expanded into its own stage by
+substituting is value in expression `${item}`. The item's value is appended to
+each stage name after a `@`. `dvc.lock` will reflect this:
+
+```yaml
+stages:
+  echo@bar:
+    cmd: echo "bar"
+  echo@baz:
+    cmd: echo "baz qux"
+  echo@foo:
+    cmd: echo "foo"
+```
+
+For lists containing complex values (e.g. dictionaries), the substitution
+expression can use the `${item.key}` form. Stage names will be appended with a
+zero-based index. For example:
+
+```yaml
+stages:
+  train:
+    foreach:
+      - epochs: 3
+        thresh: 10
+      - epochs: 10
+        thresh: 15
+    cmd: python train.py ${item.epochs} ${item.thresh}
+```
+
+```yaml
+# dvc.lock
+stages:
+  train@0:
+    cmd: python train.py 3 10
+  train@1:
+    cmd: python train.py 10 15
+```
+
+DVC can also iterate on a dictionary given directly to `foreach`, resulting in
+two substitution expressions being available: `${key}` and `${item}`. The former
+is used for the stage names:
+
+```yaml
+stages:
+  build:
+    foreach:
+      uk:
+        epochs: 3
+        thresh: 10
+      us:
+        epochs: 10
+        thresh: 15
+    do:
+      cmd: python train.py '${key}' ${item.epochs} ${item.thresh}
+      outs:
+        - model-${key}.hdfs
+```
+
+```yaml
+# dvc.lock
+stages:
+  build@uk:
+    cmd: python train.py 'uk' 3 10
+    outs:
+      - model-uk.hdfs
+  build@us: ...
+```
+
+Importantly, dictionaries from [parameters](#parameterize-dvcyaml) files can be
+used in `foreach` multi-stages as well:
+
+```yaml
+stages:
+  mystages:
+    foreach: ${myobject} # From params.yaml
+    do:
+      cmd: ./script.py ${key} ${item.prop1}
+      outs:
+        - ${item.prop2}
+```
 
-Full <abbr>parameters</abbr> (key and value) are listed separately under
-`params`, grouped by parameters file.
+> ⚠️ Remember that DVC only guarantees stage execution order based on
+> [DAGs](/doc/command-reference/dag). This means that if the resulting stages
+> are not connected by their <abbr>outputs</abbr> and <abbr>dependencies</abbr>,
+> they may not be [reproed](/doc/command-reference/repro) in the order in which
+> they are defined.
```

Comment:
Should we also insert this note in `repro` or somewhere else? (separate to this PR) It's not really mentioned anywhere.

> Related to iterative/dvc/issues/5181

Reply:
I don't find this information necessary.

Reply:
OK. I think it's a relevant warning to set expectations but we would need to mention it in other places too (out of scope for this PR). Let's see what Ivan thinks.

Reply:
I would not complicate this Guide with it. It's more about `repro` if we decide to mention it all. Let's do it separately please.

Reply:
OK sure, will extract!